src/protocols/multi_party_ecdsa/gg_2020/test.rs

```diff
@@ -150,6 +191,210 @@ fn test_sign_n5_t2_ttag4_corrupt_step7_party24() {
+//
+#[cfg(test)]
+fn keygen_orchestrator(
+    params: Parameters,
+) -> Result<
+    (
+        Vec<Keys>,
+        Vec<SharedKeys>,
+        Vec<GE>,
+        GE,
+        VerifiableSS,
+        Vec<EncryptionKey>,
+        Vec<DLogStatement>,
+    ),
+    ErrorType,
+> {
+    let participants = (0..(params.share_count as usize))
+        .map(|k| k + 1)
+        .collect::<Vec<usize>>();
+    let mut party_keys_vec = vec![];
+    let mut bc1_vec = vec![];
+    let mut decom_vec = vec![];
+    let mut h1_h2_N_tilde_vec = vec![];
+    for participant in participants.iter() {
+        let (party_keys, bc1, decom, h1_h2_N_tilde) = keygen_stage1(*participant);
+        party_keys_vec.push(party_keys);
+        bc1_vec.push(bc1);
+        decom_vec.push(decom);
+        h1_h2_N_tilde_vec.push(h1_h2_N_tilde);
+    }
+    let mut vss_scheme_vec = vec![];
+    let mut secret_shares_vec = vec![];
+    let mut index_vec = vec![];
+    for participant in participants.iter() {
+        let (vss_scheme, secret_shares, index) =
+            keygen_stage2(*participant, &params, &party_keys_vec, &bc1_vec, &decom_vec);
+        vss_scheme_vec.push(vss_scheme);
+        secret_shares_vec.push(secret_shares);
+        index_vec.push(index);
+    }
+    // The party shares are secret values.
+    // Each value in secret_shares_vec[j][i] should be encrypted by a key owned by
+    // participant i. So that those shares are only available to that participant and no
+    // one else.
+    let party_shares = (0..params.share_count)
+        .map(|i| {
+            (0..params.share_count)
+                .map(|j| {
+                    let vec_j = &secret_shares_vec[j as usize];
+                    vec_j[i as usize]
+                })
+                .collect::<Vec<FE>>()
+        })
+        .collect::<Vec<Vec<FE>>>();
+    let mut shared_keys_vec = vec![];
+    let mut dlog_proof_vec = vec![];
+    for participant in participants.iter() {
+        let (shared_keys, dlog_proof) = keygen_stage3(
+            &party_keys_vec[participant - 1],
+            &vss_scheme_vec,
+            &party_shares,
+            &decom_vec,
+            &params,
+            *participant,
+            &index_vec,
+        );
+        shared_keys_vec.push(shared_keys);
+        dlog_proof_vec.push(dlog_proof);
+    }
+    // At this point the shared_keys contain the secret values.
+    // These values should be encrypted using a key owned by that participant.
+    let pk_vec = (0..params.share_count)
+        .map(|i| dlog_proof_vec[i as usize].pk)
+        .collect::<Vec<GE>>();
+
+    let y_vec = (0..params.share_count)
+        .map(|i| decom_vec[i as usize].y_i)
+        .collect::<Vec<GE>>();
+    let mut y_vec_iter = y_vec.iter();
+    let head = y_vec_iter.next().unwrap();
+    let tail = y_vec_iter;
+    let y_sum = tail.fold(head.clone(), |acc, x| acc + x);
+
+    let dlog_verification = Keys::verify_dlog_proofs(&params, &dlog_proof_vec, &y_vec);
```

Comment:
This test must be done locally by each party

Reply:
added stage4 for that.